mastodon/Mastodon.py

```diff
@@ -1017,19 +1035,39 @@ def __api_request(self, method, endpoint, params = {}, files = {}, do_ratelimiti
 
         return response
 
-    def __stream(self, endpoint, listener, params = {}):
+
+    def __stream(self, endpoint, listener, params = {}, async=False):
         """
         Internal streaming API helper.
+
+        Returns the requests.Response instance corresponding to the open websocket
+        connection.
         """
 
         headers = {}
         if self.access_token != None:
             headers = {'Authorization': 'Bearer ' + self.access_token}
-
         url = self.api_base_url + endpoint
-        with closing(requests.get(url, headers = headers, data = params, stream = True)) as r:
-            listener.handle_stream(r.iter_lines())
 
+        connection = requests.get(url, headers = headers, data = params, stream = True)
+
+        def __stream_threadproc():
+            with closing(connection) as r:
+                try:
+                    listener.handle_stream(r.iter_lines())
+                except AttributeError as e:
+                    # TODO If the user closes the connection early, requests gets
+                    # confused and throws an AttributeError
+                    pass
+            return 0
+
+        if async:
+            t = threading.Thread(args=(), target=__stream_threadproc)
+            t.start()
+            return connection
```

Comment:
fundamentally, it feels unsafe that you're passing the whole Response object back. since the consumer could, for example, call `iter_lines`, which would be a huge problem. I'm also not sure how thread-safe `.close` is supposed to be—which could be the attribute error you're getting.

However, given the current way the code is split up, it feels difficult to put in a synchronized variable you could use the close the connection. Maybe there should be a `listener.close()` method instead? Have you considered just killing the thread?